### DISCUSSION TITLE
When quiz time exceeds logout for inactivity

### DISCUSSION THREAD

**Discussion by Alex-Jordan** [openwebwork/webwork2#1409] [Q&A]
An issue that happens at my institution goes something like this. A novice instructor has their course's logout-for-inactivity time just being our site default of 2.5 hours. They create a midterm or a final exam as a Gateway quiz, put all the questions on one page, and set the quiz time for 3 hours. Then a student starts the quiz. The student is actively working on math exercises, even typing in some answers, but not clicking anything until close to the 3 hour cap. When the student finally clicks to submit, they are logged out for inactivity. The answers were never recorded anywhere. The course is set up so that students must enter through the LMS, so they can't simply type in a password then and there to let the form data go through.  Shen they get back in to WeBWorK, the answers they had typed are gone.

So that's the problem, and my question is what strategy to pursue to do something about it. These things I've thought of all have pros and cons.

1. While a timed quiz is active for a student, it could be arranged so that you do not get logged out for inactivity when clicking something from a Gateway page.
2. When the instructor is setting a time limit for the quiz, there could be a reminder to make it shorter than the inactivity time.
3. When the instructor is setting a time limit for the quiz, it could throw an error if they set a time that is longer than the inactivity time.
4. There can be the inactivity time set in the configuration chain, but that is not really what is used. What is used is the maximum of that time and all quiz time limits (plus a few minutes given the quiz grace period).
5. Even though the student is being signed out, the answers in their form data could still be saved.

Do any of these strategies sound the least bad? Or can anyone imagine something better?

This happens at least once a term at my school, where we have many instructors and significant turnover. Educating instructors (to increase their inactivity time) helps, but it would be nice to do something more.

**Comment by mgage** [comment]
I'm wondering about your 5th suggestion.  While it's been a long time since I worked on the re-login feature it seems to me that if you were not using LTI the student would be given a re-login page and hidden in that page would be stored all of the information that had been collected and stored in answers/hidden answers worked on by the student.  If the student successfully logs in that page would be restored. With LTI all of that information is lost.  If the information was stored either in local storage or on the server then it might be possible to reclaim it after the LTI mechanism has re-logged in the student. This feature would be useful for ordinary homework problems although the consequences of loosing information are not as bad there.  Local storage would seem to be a good match for this since it seems reasonable to expect that a student would immediately log in again, on the same machine, after being timed out.

**Comment by dlglin** [comment]
In the long run the best solution would be that answers are regularly saved without a student having to submit the entire form, as this would solve several problems, but that is a much bigger ask.

In my case I have set the session timeout for the server to be longer than the longest exams we typically run.  I can't remember an instance where the long sessions caused any real problems.  I guess it's technically bad for security, but I've never run into an actual security issue because of it.

I will say that I am opposed to option 4.  I think that having the system be hard-coded to override user settings leads to bad user experiences.  I can rationalize option 1 because the student is technically still "active", even if the server can't detect it.  The one concern I could foresee with option 1 is as follows:
A student is working on a quiz with a 24 hour time limit on a public computer.  They decide to leave and work on the quiz later, but forget to log out.  Under this scenario they will remain logged in on that computer for the full 24 hours (or until they log in somewhere else).

**Comment by ionparticle** [comment]
We added a hack that uses jquery to do [a preview submit in the background](https://github.com/ubc/webwork2/blob/f959da01a28c3ed46f6fafbf663da8895df3e772/lib/WeBWorK/ContentGenerator/GatewayQuiz.pm#L2264-L2294) every 3-4 minutes. But we still get students who'd lose their answers because they logged in to Webwork on another browser and invalidated their session (e.g.: swapping between a laptop and an ipad). So the autosave alone isn't enough, it probably also should have some kind of session checking to warn students when their session has expired.
